From bdb64e5d6c85349ac54c06576b444945fa4f653f Mon Sep 17 00:00:00 2001
From: Massimo Pennazio <maxipenna@libero.it>
Date: Fri, 8 Nov 2024 11:10:32 +0000
Subject: [PATCH 1/4] Comment out drop_capabilities_bounding_set_if_needed
Upstream-Status: Pending
---
 adb/adb.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/adb/adb.c b/adb/adb.c
index 10a1e0da26..1aa5acfee5 100644
--- a/adb/adb.c
+++ b/adb/adb.c
@@ -1365,8 +1365,9 @@ int adb_main(int is_daemon, int server_port)
     /* don't listen on a port (default 5037) if running in secure mode */
     /* don't run as root if we are running in secure mode */
     if (should_drop_privileges()) {
+#if 0
         drop_capabilities_bounding_set_if_needed();
-
+#endif
         /* then switch user and group to "shell" */
         if (setgid(AID_SHELL) != 0) {
             exit(1);
-- 
2.25.1

