From 54fca1a61d40749d5865b2d1eefe566647c0a0c9 Mon Sep 17 00:00:00 2001
From: Massimo Pennazio <maxipenna@libero.it>
Date: Fri, 8 Nov 2024 11:19:15 +0000
Subject: [PATCH 4/4] Modified shell command to /bin/bash
Upstream-Status: Pending

---
 adb/services.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/adb/services.c b/adb/services.c
index d44b0c5068..7c989ecfd6 100644
--- a/adb/services.c
+++ b/adb/services.c
@@ -299,7 +299,7 @@ static int create_subproc_raw(const char *cmd, const char *arg0, const char *arg
 }
 #endif  /* !ABD_HOST */
 
-#define SHELL_COMMAND "/bin/sh"
+#define SHELL_COMMAND "/bin/bash"
 
 #if !ADB_HOST
 static void subproc_waiter_service(int fd, void *cookie)
-- 
2.25.1

