diff --git a/drivers/mmc/fsl_esdhc_imx.c b/drivers/mmc/fsl_esdhc_imx.c
index 65dee28fdc..31399062d7 100644
--- a/drivers/mmc/fsl_esdhc_imx.c
+++ b/drivers/mmc/fsl_esdhc_imx.c
@@ -1595,6 +1595,14 @@ static int fsl_esdhc_probe(struct udevice *dev)
 	}
 #endif
 
+	/* WORKAROUND: Force 200MHz for MW2816 eMMC on usdhc3 */
+	if ((ulong)priv->esdhc_regs == 0x30b60000) {
+			printf("*** USDHC3: Forcing sdhc_clk from %uHz to 200MHz for MW2816 eMMC ***\n",
+				   priv->sdhc_clk);
+			priv->sdhc_clk = 200000000;
+	}
+	printf("USDHC@0x%lx: sdhc_clk=%uHz\n", (ulong)priv->esdhc_regs, priv->sdhc_clk);
+
 	ret = fsl_esdhc_init(priv, plat);
 	if (ret) {
 		dev_err(dev, "fsl_esdhc_init failure\n");
