commit 875e02cf631407173595c10b3a9bddf9b5629011
Author: Massimo Pennazio <maxipenna@libero.it>
Date:   Fri May 31 10:06:37 2024 +0200

    Test: move back console to lpuart1

diff --git a/recipes-bsp/u-boot/u-boot-imx/portenta-x9/portenta-x9.c b/recipes-bsp/u-boot/u-boot-imx/portenta-x9/portenta-x9.c
index 1946b9b..db49ccb 100644
--- a/recipes-bsp/u-boot/u-boot-imx/portenta-x9/portenta-x9.c
+++ b/recipes-bsp/u-boot/u-boot-imx/portenta-x9/portenta-x9.c
@@ -23,19 +23,19 @@
 
 DECLARE_GLOBAL_DATA_PTR;
 
-#define UART_PAD_CTRL	(PAD_CTL_DSE(6) | PAD_CTL_FSEL2 | PAD_CTL_PUE)
-#define WDOG_PAD_CTRL	(PAD_CTL_DSE(6) | PAD_CTL_ODE | PAD_CTL_PUE | PAD_CTL_PDE)
+#define UART_PAD_CTRL	(PAD_CTL_DSE(6) | PAD_CTL_FSEL2)
+#define WDOG_PAD_CTRL	(PAD_CTL_DSE(6) | PAD_CTL_ODE | PAD_CTL_PUE | PAD_CTL_PE)
 
 static iomux_v3_cfg_t const uart_pads[] = {
-	MX93_PAD_GPIO_IO05__LPUART6_RX | MUX_PAD_CTRL(UART_PAD_CTRL),
-	MX93_PAD_GPIO_IO04__LPUART6_TX | MUX_PAD_CTRL(UART_PAD_CTRL),
+	MX93_PAD_UART1_RXD__LPUART1_RX | MUX_PAD_CTRL(UART_PAD_CTRL),
+	MX93_PAD_UART1_TXD__LPUART1_TX | MUX_PAD_CTRL(UART_PAD_CTRL),
 };
 
 int board_early_init_f(void)
 {
 	imx_iomux_v3_setup_multiple_pads(uart_pads, ARRAY_SIZE(uart_pads));
 
-	init_uart_clk(LPUART6_CLK_ROOT);
+	init_uart_clk(LPUART1_CLK_ROOT);
 
 	return 0;
 }
diff --git a/recipes-bsp/u-boot/u-boot-imx/portenta-x9/portenta-x9.dts b/recipes-bsp/u-boot/u-boot-imx/portenta-x9/portenta-x9.dts
index 3157463..79064ef 100644
--- a/recipes-bsp/u-boot/u-boot-imx/portenta-x9/portenta-x9.dts
+++ b/recipes-bsp/u-boot/u-boot-imx/portenta-x9/portenta-x9.dts
@@ -12,8 +12,7 @@
 	compatible = "fsl,imx93-9x9-qsb", "fsl,imx93";
 
 	chosen {
-		bootargs = "console=ttyLP5,115200";
-		stdout-path = &lpuart6;
+		stdout-path = &lpuart1;
 	};
 
 	reserved-memory {
@@ -214,10 +213,10 @@
 	status = "okay";
 };
 
-&lpuart1 {
+&lpuart1 { /* console */
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_uart1>;
-	status = "disabled";
+	status = "okay";
 };
 
 &lpuart2 {
@@ -226,7 +225,7 @@
 	status = "disabled";
 };
 
-&lpuart6 { /* console */
+&lpuart6 {
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_uart6>;
 	clocks = <&clk IMX93_CLK_LPUART6_GATE>, <&clk IMX93_CLK_LPUART6_GATE>;
@@ -234,6 +233,14 @@
 	status = "okay";
 };
 
+&mu1 {
+	status = "okay";
+};
+
+&mu2 {
+	status = "okay";
+};
+
 &usbotg1 { /* Type-C USB Port USB_HS */
 	dr_mode = "peripheral";
 	hnp-disable;
@@ -245,14 +252,6 @@
 	status = "okay";
 };
 
-&mu1 {
-	status = "okay";
-};
-
-&mu2 {
-	status = "okay";
-};
-
 &usbotg2 { /* USB_FS */
 	dr_mode = "host";
 	hnp-disable;
diff --git a/recipes-bsp/u-boot/u-boot-imx/portenta-x9/portenta-x9.h b/recipes-bsp/u-boot/u-boot-imx/portenta-x9/portenta-x9.h
index 896281e..8ca592b 100644
--- a/recipes-bsp/u-boot/u-boot-imx/portenta-x9/portenta-x9.h
+++ b/recipes-bsp/u-boot/u-boot-imx/portenta-x9/portenta-x9.h
@@ -48,7 +48,7 @@
 	"kernel_addr_r=" __stringify(CONFIG_SYS_LOAD_ADDR) "\0" \
 	"image=Image\0" \
 	"splashimage=0x90000000\0" \
-	"console=ttyLP5,115200 earlycon\0" \
+	"console=ttyLP0,115200 earlycon\0" \
 	"fdt_addr_r=0x83000000\0"			\
 	"fdt_addr=0x83000000\0"			\
 	"fdt_high=0xffffffffffffffff\0"		\
