From e2f9df212ef4df9c4403bc2183ce237978d92e56 Mon Sep 17 00:00:00 2001
From: maidnl <d.aimo@arduino.cc>
Date: Tue, 2 Sep 2025 14:54:08 +0200
Subject: [PATCH] Enable tcp with env variable

---
 adb/adb.c | 17 +++++++++++++----
 1 file changed, 13 insertions(+), 4 deletions(-)

diff --git a/adb/adb.c b/adb/adb.c
index 10a1e0da2..b1ac57841 100644
--- a/adb/adb.c
+++ b/adb/adb.c
@@ -1400,11 +1400,20 @@ int adb_main(int is_daemon, int server_port)
     // If one of these properties is set, also listen on that port
     // If one of the properties isn't set and we couldn't listen on usb,
     // listen on the default port.
-    property_get("service.adb.tcp.port", value, "");
-    if (!value[0]) {
-        property_get("persist.adb.tcp.port", value, "");
+    //
+    port = 0; // do not use port if ENV variable is not set
+    const char* env_var_name = "SERVICE_ADB_TCP_PORT";
+    char* env_var_value_str = getenv(env_var_name);
+    if (env_var_value_str != NULL) {
+        char* endptr;
+        long int value = strtol(env_var_value_str, &endptr, 10);
+        port = (int)value;
+        if(port < 0 || port > 65535) {
+            port = 0; //do not use port with wrong values
+        }
     }
-    if (sscanf(value, "%d", &port) == 1 && port > 0) {
+
+    if (port > 0) {
         printf("using port=%d\n", port);
         // listen on TCP port specified by service.adb.tcp.port property
         local_init(port);
-- 
2.43.0

